Implement deactivating machines (instead of deleting)

dancek opened the pull request:
Implementation of feature request #3. This is not yet ready for merge. I'm willing to fix things as needed.

The functionality is basically one boolean field in the `machines` database table, a button+controller for setting it to `false` and filtering the machine list to only show active machines.

My questions and known issues are:

- [x] Firstly this really needs code review. It's the first time I've read or written Elixir and there might be some extremely stupid code.
- [x] Are machines displayed elsewhere than in `/my/machines`, ie. should the flag be checked elsewhere? Was it a bad idea to filter only active machines in the view template?
- [x] Some functionality needs to be disabled for deactivated machines (at least saving pulses). Should this be done on a case-by-case basis, by resetting the API key or some other way?
- [x] How should the deactivate button be positioned? API key + 5 buttons on the same row seem too much to me. I didn't know what to do so did nothing at this point.
- [x] There's no confirmation like for delete (I just remembered).
- [x] Should it be possible to reactivate a machine?

dancek committed:
Implement deactivating machines (instead of deleting)

Resolves #3.

Nicd commented:
Thank for the PR! I will add some comments and more later when I have more time.

Nicd commented on web/controllers/machine_controller.ex:
```diff
@@ -75,6 +75,18 @@ defmodule CodeStats.MachineController do
+        |> put_flash(:success, "Machine #{machine.name} deactivated.")
+        |> redirect(to: machine_path(conn, :list))
+      end
+  end
```
Good use of with. The `do` I'd like to be on its own line so that the indentation stays saner but I'll have to check what the official formatter coming in 1.6 says about that.

Nicd replied:
This needs handling for the failure cases (only the happy path is implemented).

dancek replied:
I thought `edit_machine_or_flash` handles the errors. Isn't that enough?

Nicd replied:
Ah yes indeed, I have been cleverer than myself when I wrote that. Probably not the clearest of codes. Should look into how to rewrite it with https://hexdocs.pm/phoenix/Phoenix.Controller.html#action_fallback/1 later but this will do for now.

Nicd commented on web/models/machine.ex:
```diff
@@ -22,11 +23,12 @@ defmodule CodeStats.Machine do
   def update_changeset(data, params \\ %{}) do
     data
-    |> cast(params, [:name])
+    |> cast(params, [:name, :active])
```
It might be better to have a separate changeset for updating the active status. Now this changeset requires both `:name` and `:active` and that may lead to problems and inadvertent changes of the other value when the other is changed and the other is left unset.

Nicd commented on web/templates/machine/machines.html.eex:
```diff
@@ -2,7 +2,7 @@
 
   <div class="machine-list">
-    <%= for machine <- @machines do %>
+    <%= for %{active: true} = machine <- @machines do %>
```
Nice trick! I will have to keep this in mind.

Nicd replied:
But this list should show all machines so that they can be reactivated again. Or maybe deactivated machines can be shown in a separate list below with only action "Activate"?

dancek replied:
Yeah, I thought about that. The issue said:
> Deactivation should make a machine invisible in the machines list

I think it makes sense to add a second list in the bottom.

Nicd replied:
Yeah the comment is unfortunately ambiguous. It meant that the machine should be made invisible in the _profile page_ machines list. I should have worded it better. :)

Nicd commented:
> Firstly this really needs code review. It's the first time I've read or written Elixir and there might be some extremely stupid code.

It was pretty good for a first timer! 👍 

>  Are machines displayed elsewhere than in /my/machines, ie. should the flag be checked elsewhere? Was it a bad idea to filter only active machines in the view template?

Machines will be displayed on the profile page in some widgets. Take a look at the bottom of the current profile page: https://codestats.net/users/Nicd

Filtering like that is not a bad idea if we have a separate list for deactivated machines.

>  Some functionality needs to be disabled for deactivated machines (at least saving pulses). Should this be done on a case-by-case basis, by resetting the API key or some other way?

Pulses should not be accepted for deactivated machines. The API key doesn't need to be reset IMO, in AuthUtils's `defp get_machine` we can add condition `active == true` and then it won't find inactive machines anymore. Might be needed to check that the editor plugins don't hold on to the XP and try to resend forever though, so maybe we need to use a special HTTP code to signify that (404?).

>  How should the deactivate button be positioned? API key + 5 buttons on the same row seem too much to me. I didn't know what to do so did nothing at this point.

Yeah it becomes a mess. Maybe the buttons can be below the API key at all screen sizes? I'm not a good UI designer. :D

>  There's no confirmation like for delete (I just remembered).

That would be nice even though deactivation is not a destructive feature.

>  Should it be possible to reactivate a machine?

Yes, that needs to be added. The point of deactivation is that it's not permanent and you don't lose your XP like with deletion.

Nicd commented on web/static/frontend/css/machines-layout.scss:
```diff
@@ -64,3 +63,7 @@
+
+.inactive-machine-list {
+  // TODO
+}
```
Files should end with a trailing newline

[Read more about it here.](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#finalnewline)

Nicd commented on web/static/frontend/css/machines-layout.scss:
```diff
@@ -64,3 +63,7 @@
 }
+
+.inactive-machine-list {
+  // TODO
```
TODO found

Nicd commented on web/controllers/machine_controller.ex:
```diff
@@ -96,6 +96,28 @@ defmodule CodeStats.MachineController do
+  defp activate_or_deactivate(conn, id, active) do
+    user = SetSessionUser.get_user_data(conn)
+    verb = active && "activated" || "deactivated"
+    
```
There should be no trailing white-space at the end of a line.

Nicd commented on web/models/machine.ex:
```diff
@@ -30,6 +32,11 @@ defmodule CodeStats.Machine do
 
+  def activation_changeset(data, params \\ %{}) do
+    data
+    |> cast(params, [:active])
```
Use a function call when a pipeline is only one function long

Nicd commented on web/utils/profile_utils.ex:
```diff
@@ -66,7 +66,9 @@ defmodule CodeStats.ProfileUtils do
       join: p in Pulse, on: m.id == p.machine_id,
       join: x in XP, on: p.id == x.pulse_id,
-      where: m.user_id == ^user.id and p.sent_at >= ^then,
+      where: m.user_id == ^user.id and 
```
There should be no trailing white-space at the end of a line.

Nicd commented on web/utils/profile_utils.ex:
```diff
@@ -100,7 +102,8 @@ defmodule CodeStats.ProfileUtils do
 
     machine_q = from m in Machine,
-      where: m.user_id == ^user.id,
+      where: m.user_id == ^user.id and 
```
There should be no trailing white-space at the end of a line.

Nicd commented:
Ebert has finished reviewing this Pull Request and has found:

* 6 possible new issues (including those that may have been commented here).
* 1 fixed issue! 🎉

You can see more details about this review at https://ebertapp.io/github/code-stats/code-stats/pulls/45.

<!-- Ebert auto generated comment -->

Nicd commented on web/static/frontend/css/machines-layout.scss:
```diff
@@ -64,3 +63,12 @@
+.inactive-machine-list {
+  >.machine {
+    display: grid;
+    grid-template-columns: 1fr auto;
```
fr units not allowed on `grid-template-columns`; must be one of (%, Hz, ch, cm, deg, dpcm, dpi, dppx, em, ex, grad, in, kHz, mm, ms, pc, pt, px, q, rad, rem, s, turn, vh, vmax, vmin, vw)

Nicd commented on web/static/frontend/css/machines-layout.scss:
```diff
@@ -8,21 +8,16 @@
-      grid-template: 'name   name   name   name   name  '
-                     'key    copy   regen  edit   delete'
-                    / 1fr    auto   auto   auto   auto;
+      grid-template: 'name   name   name   name  '
```
fr units not allowed on `grid-template`; must be one of (%, Hz, ch, cm, deg, dpcm, dpi, dppx, em, ex, grad, in, kHz, mm, ms, pc, pt, px, q, rad, rem, s, turn, vh, vmax, vmin, vw)

Nicd replied:
fr units not allowed on `grid-template`; must be one of (%, Hz, ch, cm, deg, dpcm, dpi, dppx, em, ex, grad, in, kHz, mm, ms, pc, pt, px, q, rad, rem, s, turn, vh, vmax, vmin, vw)

Nicd replied:
fr units not allowed on `grid-template`; must be one of (%, Hz, ch, cm, deg, dpcm, dpi, dppx, em, ex, grad, in, kHz, mm, ms, pc, pt, px, q, rad, rem, s, turn, vh, vmax, vmin, vw)

Nicd replied:
fr units not allowed on `grid-template`; must be one of (%, Hz, ch, cm, deg, dpcm, dpi, dppx, em, ex, grad, in, kHz, mm, ms, pc, pt, px, q, rad, rem, s, turn, vh, vmax, vmin, vw)

Nicd commented on web/static/frontend/js/machine_page.js:
```diff
@@ -4,14 +4,20 @@
 
+    const deactivate_buttons = document.getElementsByClassName('machine-deactivate-button');
+
+    for (const button of deactivate_buttons) {
```
Identifier 'deactivate_buttons' is not in camel case.

Enforce the usage of camel case to have a consistent codebase.

[Read more about it here.](http://eslint.org/docs/rules/camelcase)

Nicd commented on web/static/frontend/js/machine_page.js:
```diff
@@ -4,14 +4,20 @@
       };
     }
 
+    const deactivate_buttons = document.getElementsByClassName('machine-deactivate-button');
```
Identifier 'deactivate_buttons' is not in camel case.

Enforce the usage of camel case to have a consistent codebase.

[Read more about it here.](http://eslint.org/docs/rules/camelcase)

Nicd commented on web/static/frontend/js/machine_page.js:
```diff
@@ -4,14 +4,20 @@
+    const delete_buttons = document.getElementsByClassName('machine-delete-button');
 
-    for (const button of buttons) {
+    for (const button of delete_buttons) {
```
Identifier 'delete_buttons' is not in camel case.

Enforce the usage of camel case to have a consistent codebase.

[Read more about it here.](http://eslint.org/docs/rules/camelcase)

Nicd commented on web/static/frontend/js/machine_page.js:
```diff
@@ -4,14 +4,20 @@
 function machine_page() {
   document.addEventListener('DOMContentLoaded', () => {
-    const buttons = document.getElementsByClassName('machine-delete-button');
+    const delete_buttons = document.getElementsByClassName('machine-delete-button');
```
Identifier 'delete_buttons' is not in camel case.

Enforce the usage of camel case to have a consistent codebase.

[Read more about it here.](http://eslint.org/docs/rules/camelcase)

Nicd commented:
Ebert has finished reviewing this Pull Request and has found:

* 10 possible new issues (including those that may have been commented here).
* 1 fixed issue! 🎉

You can see more details about this review at https://ebertapp.io/github/code-stats/code-stats/pulls/45.

<!-- Ebert auto generated comment -->

dancek commented:
This is probably ready for merge after I rebase to current `grid-refactor` and squash these commits. I think I fixed all relevant issues and implemented discussed changes (hope I didn't miss anything). Of course Ebert still complains, but the poor thing is wrong.

dancek commented:
And there it is, squashed.

dancek commented:
Oh, one more thing. I didn't really get to test the profile page as the frontend doesn't seem to display machines currently. I hope the filtering in server side is done correctly.

Nicd commented:
Thanks @dancek <3
